File: src/main/java/org/distril/beengine/world/chunk/ChunkManager.java
package org.distril.beengine.world.chunk;

import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.distril.beengine.util.AsyncArrayValue;
import org.distril.beengine.util.ChunkUtils;
import org.distril.beengine.world.World;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ThreadLocalRandom;

@Log4j2
@RequiredArgsConstructor
public class ChunkManager {

	private final Map<Long, Chunk> chunks = new ConcurrentHashMap<>();

	private final World world;

	public Chunk getLoadedChunk(int x, int z) {
		return this.getLoadedChunk(ChunkUtils.key(x, z));
	}

	public Chunk getLoadedChunk(long key) {
		return this.chunks.get(key);
	}

	public Chunk getChunk(int x, int z) {
		return this.getChunk(ChunkUtils.key(x, z));
	}

	public Chunk getChunk(long key) {
		var chunk = this.getLoadedChunk(key);
		if (chunk == null) {
			chunk = this.generateChunk(key).join();
		}

		return chunk;
	}

	public CompletableFuture<Chunk> generateChunk(int x, int z) {
		return this.generateChunk(ChunkUtils.key(x, z));
	}

	public CompletableFuture<Chunk> generateChunk(long key) {
		return CompletableFuture.supplyAsync(() -> {
			var loadedChunk = this.getLoadedChunk(key);
			if (loadedChunk != null) {
				return loadedChunk;
			}

			var chunk = this.chunks.computeIfAbsent(key, chunkKey -> new Chunk(ChunkUtils.fromKeyX(chunkKey),
					ChunkUtils.fromKeyZ(chunkKey)));

			world.getGenerator().generate(ThreadLocalRandom.current(), chunk);
			return chunk;
		});
	}

	public synchronized void tick() {
		if (this.chunks.isEmpty()) {
			return;
		}

		AsyncArrayValue<Boolean> futureArray = new AsyncArrayValue<>();
		this.chunks.forEach((key, chunk) -> futureArray.add(chunk.tick().whenComplete((close, throwable) -> {
			if (throwable != null) {
				log.error("Error when ticking chunk {}", chunk, throwable);
				return;
			}

			if (close) {
				this.unloadChunk(key, true, false);
			}
		})));

		futureArray.execute().join();
	}

	public void unloadChunk(long key, boolean save, boolean force) {
		var chunk = this.getLoadedChunk(key);
		if (chunk == null) {
			return;
		}

		if (!force && !chunk.getLoaders().isEmpty()) {
			return;
		}

		if (save) {
			// this.saveChunk(chunk);
		}

		chunk.close();

		this.chunks.remove(key);
	}

	public Map<Long, Chunk> getChunks() {
		return new HashMap<>(chunks);
	}
}

File: src/main/java/org/distril/beengine/world/chunk/ChunkManager.kt
package org.distril.beengine.world.chunk

import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flow
import org.distril.beengine.util.ChunkUtils
import org.distril.beengine.world.World
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ThreadLocalRandom

class ChunkManager(private val world: World) {

	val chunks: MutableMap<Long, Chunk> = ConcurrentHashMap()
		get() = HashMap(field)

	fun getLoadedChunk(x: Int, z: Int) = this.getLoadedChunk(ChunkUtils.encode(x, z))

	fun getLoadedChunk(key: Long) = this.chunks[key]

	fun getChunk(x: Int, z: Int) = this.getChunk(ChunkUtils.encode(x, z))

	fun getChunk(key: Long): Chunk {
		var chunk = this.getLoadedChunk(key)
		if (chunk == null) {
			chunk = this.generateChunk(key)
		}

		return chunk
	}

	fun generateChunk(x: Int, z: Int) = this.generateChunk(ChunkUtils.encode(x, z))

	fun generateChunk(key: Long): Chunk {
		val loadedChunk = this.getLoadedChunk(key)
		if (loadedChunk != null) return loadedChunk

		val chunk = this.chunks.computeIfAbsent(key) {
			Chunk(
				ChunkUtils.decodeX(it),
				ChunkUtils.decodeZ(it)
			)
		}

		this.world.generator.generate(ThreadLocalRandom.current(), chunk)
		return chunk
	}

	suspend fun tick() {
		if (this.chunks.isNotEmpty()) {
			val tickedChunks = flow {
				chunks.forEach { (key, chunk) ->
					run {
						val canBeUnload = chunk.tick()

						if (canBeUnload) unloadChunk(key)

						emit(Unit)
					}
				}
			}

			tickedChunks.collect()
		}
	}

	fun unloadChunk(key: Long, save: Boolean = true, force: Boolean = false) {
		val chunk = this.getLoadedChunk(key) ?: return
		if (!force && chunk.loaders.isNotEmpty()) return

		if (save) {
			// this.saveChunk(chunk);
		}

		chunk.close()
		this.chunks.remove(key)
	}
}
